extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/KinesisRecordSupplier.java

```diff
@@ -552,8 +552,11 @@ public void start()
     }
   }
 
+  /**
+   * This method is synchronized because multiple threads can attempt to close the record supplier concurrently
+   */
   @Override
-  public void close()
+  public synchronized void close()
```

Comment:
Maybe it is fine to leave the KinesisRecordSupplier and KafkaRecordSupplier as what it was before (without the synchronized). I think it is sufficient to fix this bug with just the change in TimedShutoffInputSourceReader. The change already make the iterator.close() from the TimedShutoffInputSourceReader idempotent, fix race/concurrent issue.

Reply:
I do think it should be fixed now, ideally, because we can't be confident that this patch actually fixes the bug without fixing the thread safety issue too.